osquery/events/linux/auditdnetlink.cpp

```diff
@@ -668,6 +675,149 @@ bool AuditdNetlink::configureAuditService() noexcept {
   return true;
 }
 
+bool AuditdNetlink::clearAuditConfiguration() noexcept {
+  int seq = audit_request_rules_list_data(audit_netlink_handle_);
+  if (seq <= 0) {
+    VLOG(1) << "Failed to list the audit rules";
+    return false;
+  }
+
+  // Attempt to list all rules
+  std::vector<AuditRuleDataObject> rule_object_list;
+
+  for (int read_retry = 0; read_retry < 3; ++read_retry) {
+    bool netlink_ready = false;
+
+    for (int poll_retry = 0; poll_retry < 3; ++poll_retry) {
+      pollfd fds[] = {{audit_netlink_handle_, POLLIN, 0}};
+
+      errno = 0;
+      int poll_status = ::poll(fds, 1, 4);
+      if (poll_status == 0) {
+        continue;
+      }
+
+      if (poll_status < 0) {
+        VLOG(1) << "pool() failed with error " << errno;
+        return false;
+      }
+
+      if ((fds[0].revents & POLLIN) != 0) {
+        netlink_ready = true;
+        break;
+      }
+    }
+
+    if (!netlink_ready) {
+      VLOG(1) << "Could not read from the audit netlink";
+      return false;
+    }
+
+    // Get the reply from the audit link
+    struct audit_reply reply = {};
+    if (audit_get_reply(
+            audit_netlink_handle_, &reply, GET_REPLY_NONBLOCKING, 0) <= 0) {
+      continue;
+    }
+
+    read_retry = 0;
+    if (reply.nlh->nlmsg_seq != static_cast<unsigned int>(seq)) {
+      continue;
+    }
+
+    if (reply.type == NLMSG_DONE) {
+      // We have finished listing the rules
+      break;
+    }
+
+    if (reply.type == NLMSG_ERROR && reply.error->error != 0) {
+      return false;
+    }
+
+    if (reply.type != AUDIT_LIST_RULES) {
+      // Skip this reply if it is not part of the rule list output
+      continue;
+    }
+
+    // Save the rule
+    auto reply_size = sizeof(struct audit_reply) + reply.ruledata->buflen;
+
+    AuditRuleDataObject reply_object;
+    reply_object.resize(reply_size);
+    if (reply_object.size() != reply_size) {
+      VLOG(1) << "Failed to read the audit rule data";
+      return false;
+    }
+
+    std::memcpy(reply_object.data(), reply.ruledata, reply_size);
+    rule_object_list.push_back(reply_object);
+  }
+
+  // Delete each rule
+  std::size_t error_count = 0;
+  for (auto& rule_object : rule_object_list) {
+    if (!deleteAuditRule(rule_object)) {
+      error_count++;
+    }
+  }
+
+  if (error_count != 0) {
+    VLOG(1) << error_count << " audit rules could not be correctly removed";
+    return false;
+  }
+
+  VLOG(1) << "The audit configuration has been cleared";
+  return true;
+}
+
+bool AuditdNetlink::deleteAuditRule(AuditRuleDataObject& rule_object) {
+  if (NLMSG_SPACE(rule_object.size()) > MAX_AUDIT_MESSAGE_LENGTH) {
+    return false;
+  }
+
+  auto* rule_data =
+      reinterpret_cast<struct audit_rule_data*>(rule_object.data());
+
+  struct audit_message request = {};
+  request.nlh.nlmsg_len = static_cast<__u32>(NLMSG_SPACE(rule_object.size()));
```

Comment:
Why use `__u32`?

Reply:
I'm using __u32 because it's the type used by the API.